src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -0,0 +1,124 @@
+
+            if (!dotnetAssemblyPath.Exists)
+            {
+                return null;
+            }
+
+            FileVersionInfo fileVersionInfo = FileVersionInfo.GetVersionInfo(dotnetAssemblyPath.FullName);
+
+            return new VisualStudioInstance(
+                name: ".NET Core SDK",
+                path: dotNetSdkPath,
+                version: new Version(fileVersionInfo.FileMajorPart, fileVersionInfo.FileMinorPart, fileVersionInfo.FileBuildPart),
+                discoveryType: DiscoveryType.DotNetSdk);
+        }
+
+        private static string hostfxr_resolve_sdk(string workingDir)
+        {
+            Process process = Process.GetCurrentProcess();
+
+            string dotNetExePath = process?.MainModule?.FileName;
+
+            if (String.IsNullOrWhiteSpace(dotNetExePath))
+            {
+                return null;
+            }
+
+            string dotNetExeDirectory = Path.GetDirectoryName(dotNetExePath);
+
+            if (String.IsNullOrWhiteSpace(dotNetExeDirectory) || !Directory.Exists(dotNetExeDirectory))
+            {
+                return null;
+            }
+
+            var buffer = new StringBuilder(capacity: 64);
+
+            for (;;)
+            {
+                int size = Interop.hostfxr_resolve_sdk(dotNetExeDirectory, workingDir, buffer, buffer.Capacity);
+                if (size <= 0)
+                {
+                    return null;
+                }
+
+                if (size <= buffer.Capacity)
+                {
+                    break;
+                }
+
+                buffer.Capacity = size;
+            }
+
+            return buffer.ToString();
+        }
+
+        private static class Interop
+        {
+            private static readonly bool RunningOnWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+
+            public static string realpath(string path)
+            {
+                var ptr = unix_realpath(path, IntPtr.Zero);
+                var result = Marshal.PtrToStringAnsi(ptr); // uses UTF8 on Unix
+                unix_free(ptr);
+                return result;
+            }
+
+            public static int hostfxr_resolve_sdk(string exeDir, string workingDir, [Out] StringBuilder buffer, int bufferSize)
+            {
+                // hostfxr string encoding is platform -specific so dispatch to the
+                // appropriately annotated P/Invoke for the current platform.
+                return RunningOnWindows
+                    ? windows_hostfxr_resolve_sdk(exeDir, workingDir, buffer, bufferSize)
+                    : unix_hostfxr_resolve_sdk(exeDir, workingDir, buffer, bufferSize);
+            }
+
+            [DllImport("hostfxr", EntryPoint = nameof(hostfxr_resolve_sdk), CharSet = CharSet.Unicode, ExactSpelling = true, CallingConvention = CallingConvention.Cdecl)]
+            private static extern int windows_hostfxr_resolve_sdk(string exeDir, string workingDir, [Out] StringBuilder buffer, int bufferSize);
```

Comment:
I have no real dog in this fight, but using `StringBuilder` as a variable output buffer has terrible performance. If this is a concern in this code path we should fix the native API, but I believe that you could probably get away with a `ref string` and have better performance with no native API change.